src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -233,179 +240,174 @@ CitusExecScan(CustomScanState *node)
+ * CitusBeginModifyScan prepares the scan state for a modification.
  *
- * The function is also smart about caching plans if the plan is local to this node.
+ * Modifications are special because:
+ * a) we evaluate function calls (e.g. nextval) here and the outcome may
+ *    determine which shards are affected by this query.
+ * b) we need to take metadata locks to make sure no write is left behind
+ *    when finalizing a shard move.
  */
 static void
-CitusBeginScanWithCoordinatorProcessing(CustomScanState *node, EState *estate, int eflags)
+CitusBeginModifyScan(CustomScanState *node, EState *estate, int eflags)
 {
 	CitusScanState *scanState = (CitusScanState *) node;
+	PlanState *planState = &(scanState->customScanState.ss.ps);
 	DistributedPlan *originalDistributedPlan = scanState->distributedPlan;
-	DistributedPlan *distributedPlan = CopyDistributedPlanWithoutCache(scanState);
-	Job *workerJob = distributedPlan->workerJob;
-	Query *jobQuery = workerJob->jobQuery;
 
-	/* we'd only get to this function with the following conditions */
-	Assert(workerJob->requiresMasterEvaluation || workerJob->deferredPruning);
+	DistributedPlan *currentPlan =
+		CopyDistributedPlanWithoutCache(originalDistributedPlan);
+	scanState->distributedPlan = currentPlan;
 
-	PlanState *planState = &(scanState->customScanState.ss.ps);
-
-	/* citus only evaluates functions for modification queries */
-	bool modifyQueryRequiresMasterEvaluation =
-		jobQuery->commandType != CMD_SELECT &&
-		(workerJob->requiresMasterEvaluation || workerJob->deferredPruning);
+	Job *workerJob = currentPlan->workerJob;
+	Query *jobQuery = workerJob->jobQuery;
+	bool evaluateAllExpressions = workerJob->requiresMasterEvaluation ||
+								  workerJob->deferredPruning;
 
-	/*
-	 * ExecuteMasterEvaluableFunctions handles both function evalation
-	 * and parameter evaluation. Pruning is most likely deferred because
-	 * there is a parameter on the distribution key. So, evaluate in both
-	 * cases.
-	 */
-	if (modifyQueryRequiresMasterEvaluation)
+	if (evaluateAllExpressions)
 	{
-		/* evaluate functions and parameters for modification queries */
+		/* evaluate both functions and parameters */
 		ExecuteMasterEvaluableFunctionsAndParameters(jobQuery, planState);
-	}
-	else if (jobQuery->commandType == CMD_SELECT && !workerJob->deferredPruning)
-	{
-		/* we'll use generated strings, no need to have the parameters anymore */
-		EState *executorState = planState->state;
-		ResetExecutionParameters(executorState);
 
-		/* we're done, we don't want to evaluate functions for SELECT queries */
-		return;
+		/* job query no longer has parameters, so we should not send any */
+		workerJob->parametersInJobQueryResolved = true;
+
+		if (!workerJob->deferredPruning)
+		{
+			/*
+			 * After evaluating functions we need to rebuild query string.
+			 * In case of deferred pruning this happens in the Regenerate
+			 * functions.
+			 */
+			RebuildQueryStrings(jobQuery, workerJob->taskList);
```

Comment:
The only complex piece  that I find in this PR is here. There is no clear reason for readers why `RebuildQueryStrings ` does not consider `workerJob->parametersInJobQueryResolved` which is set just one line above. As there seems no harm, maybe we use it in `RebuildQueryStrings ` to update tasks?

In the same context, the readers wouldn't have a good understanding on why `RegenerateTaskListForInsert / RegenerateTaskForFasthPathQuery ` is different than `RebuildQueryStrings `, as the names are somehow close to each other.

Some more comment might be useful, or even getting the name change back for `HandleDeferredShardPruningForXXX `